Phase2 GMT Emulator

zhenbinwu opened the pull request:
#### PR description:

This is the latest version of the Phase2 GMT emulator. It contains several functions:

- The streaming track from L1TrackTrigger was converted locally
- The track are matched to muon to produce trackMuons
- The muon isolation is calculated
- The standalone muons from muon track finders are produced for Correlator layer1
- The tau to 3mu module is still under development. Currently skip in the production.

#### PR validation:

The emulators are matched to the firmware output.

cecilecaillol commented on DataFormats/L1TMuonPhase2/src/SAMuon.cc:
```diff
@@ -0,0 +1,26 @@
+SAMuon::~SAMuon() {}
+
+void SAMuon::print() const {
+  printf("Standalone Muon: charge=%d pt=%d,%f eta=%d,%f phi=%d,%f z0=%d d0=%d isolation=%d beta=%d quality=%d\n",
```
Could you please replace all printf by LogDebug (or LogWarning, LogError, etc)? https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger#LogDebug

cecilecaillol commented on DataFormats/L1TMuonPhase2/src/TrackerMuon.cc:
```diff
@@ -0,0 +1,38 @@
+TrackerMuon::~TrackerMuon() {}
+
+void TrackerMuon::print() const {
+  printf("Tracker Muon: charge=%d pt=%d,%f eta=%d,%f phi=%d,%f z0=%d d0=%d isolation=%d beta=%d quality=%d\n",
```
LogDebug

cecilecaillol commented on L1Trigger/Phase2L1GMT/interface/ConvertedTTTrack.h:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    void printWord() const {
+      printf(" %08llx%016llx",
```
LogDebug

cecilecaillol commented on L1Trigger/Phase2L1GMT/interface/Isolation.h:
```diff
@@ -0,0 +1,243 @@
+    int iso = (accum <= absiso_thrT ? 3 : accum <= absiso_thrM ? 2 : accum <= absiso_thrL ? 1 : 0);
+
+    if (verbose_) {
+      cout << " [DEBUG Isolation] : absiso_threshold L : " << absiso_thrL << " accum " << accum
+           << " bit set : " << (accum < absiso_thrL) << endl;
+      cout << " [DEBUG Isolation] : absiso_threshold M : " << absiso_thrM << " accum " << accum
+           << " bit set : " << (accum < absiso_thrM) << endl;
+      cout << " [DEBUG Isolation] : absiso_threshold T : " << absiso_thrT << " accum " << accum
+           << " bit set : " << (accum < absiso_thrT) << endl;
+      cout << " [DEBUG Isolation] : absiso : " << (iso) << endl;
```
LogDebug

cecilecaillol commented on L1Trigger/Phase2L1GMT/interface/Isolation.h:
```diff
@@ -0,0 +1,243 @@
+
+    int iso = (accum <= thrT.to_int() ? 3 : accum <= thrM.to_int() ? 2 : accum <= thrL.to_int() ? 1 : 0);
+
+    if (verbose_) {
+      cout << " [DEBUG Isolation] : reliso_threshold L : " << thrL << " accum " << accum
+           << " bit set : " << (accum < thrL.to_int()) << endl;
+      cout << " [DEBUG Isolation] : reliso_threshold M : " << thrM << " accum " << accum
+           << " bit set : " << (accum < thrM.to_int()) << endl;
+      cout << " [DEBUG Isolation] : reliso_threshold T : " << thrT << " accum " << accum
+           << " bit set : " << (accum < thrT.to_int()) << endl;
+      cout << " [DEBUG Isolation] : reliso : " << (iso << 2) << " org " << iso << endl;
+    }
```
LogDebug

gpetruc commented on L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTSAMuonProducer.cc:
```diff
@@ -0,0 +1,196 @@
+  word = word.concat(pt);
+  word = word.concat(ap_uint<1>(charge));
+  word = word.concat(qual);
+  //word = word.concat(bx);
```
this is different from what we had discussed before and put in the correlator interface document: BITSETA was 12, while here you have 13, and since you've removed the bx all the fields are shifted.

should we update the interface document?

gpetruc replied:
also, this piece of code in lines 159-166 doesn't seem to compile, at least in `CMSSW_11_1_7`

gpetruc commented on L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTSAMuonProducer.cc:
```diff
@@ -0,0 +1,196 @@
+  ap_uint<BITSQUALITY> qual = mu.hwQual();
+  // We might not need to send BX info
+  //ap_int<BITSBX> bx(bx_);
+  bool charge(mu.charge() > 0);
```
Also this is inconsistent with the [interface document](https://www.overleaf.com/read/ptkfbjyyxcvv), that has the charge bit defined as '0 = positive, 1 = negative'.
I would be happy with this new definition, as it would match with what we're using in the correlator, but you should at least inform us when you're changing things, as we need to propagate those changes to our firmware.

gpetruc commented on L1Trigger/Phase2L1GMT/python/gmt_cfi.py:
```diff
@@ -0,0 +1,82 @@
+)
+
+standaloneMuons = cms.EDProducer('Phase2L1TGMTSAMuonProducer',
+                                 muonToken  = cms.untracked.InputTag('simGmtStage2Digis', "", "HLT"),
```
do you really want to hard-code the process name here? why?

gpetruc replied:
also, an **untracked** input tag?

zhenbinwu commented:
@cecilecaillol  Thanks for the suggestions and the twiki.

I made the modification according to the twiki, moving the print() to LogDebug.
For some of the print out that are already controlled by the verbose option, which can be set via config file, I think they are common practices and kept them as they are.

zhenbinwu commented:
@gpetruc Thank you for your comments

- We changed to 13 bits for Eta for consistency in GMT. We will update the interface document.
- I corrected the charge bug. It is  '0 = positive, 1 = negative'. 
- The cfi file is also updated.
- For the compilation, the issue comes from the hls code. As you experienced, there are many warnings from hls during compilation. In CMSSW_11_3_X, the hls will compiled with -iSystem, which will ignore those warnings. I also made a local fix to ignore those warnings and missed the issue. Anyway, I have updated the code to use a local method.

rekovic commented:
Hi @zhenbinwu 
If this PR is ready now, including solution to compilation problems, can you please remove the draft tag from it.

cecilecaillol commented:
@rekovic Let's merge this?

zhenbinwu commented:
Syncing this PR with the l1t-phase2-v3.3.32, I run into the same issue as reported in https://github.com/cms-l1t-offline/cmssw/pull/901#issuecomment-810211580 when running over Phase2HLTTDRWinter20DIG samples, with Extended2026D49 geometry. 

@rekovic What was the solution to the issue you reported?   Thanks

rekovic commented:
Hi @zhenbinwu. 
I am still seeing the error of  https://github.com/cms-l1t-offline/cmssw/pull/901#issuecomment-810211580, even after the latest-greatest tags (attn @aperloff)

zhenbinwu commented:
@rekovic Thanks. We found that we need to run "git cms-addpkg L1Trigger/TrackTrigger" and rescram. It will fix the geometry issue above, though we don't know why.

aperloff commented:
> Hi @zhenbinwu.
> I am still seeing the error of [#901 (comment)](https://github.com/cms-l1t-offline/cmssw/pull/901#issuecomment-810211580), even after the latest-greatest tags (attn @aperloff)

As @zhenbinwu said, this extra package is needed when working with tracks and is already documented [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideL1TPhase2Instructions#Run_TrackTrigger_and_L1T_on_HLT). The reason this is needed is because some packages are dependent on modules that we change, but weren't directly modified by changes in cms-l1t-offline. That means that when the cms-l1t-offline changes are merged in, not all of the dependent packages get checked out and recompiled. This won't be a problem when things for merged into cms-sw/cmssw as it is built completely twice a day.

zhenbinwu commented:
@cecilecaillol @rekovic Can we merge this PR given the issue is solved? Thanks

gpetruc commented on L1Trigger/Phase2L1GMT/interface/Constants.h:
```diff
@@ -0,0 +1,1293 @@
+  }
+
+  template <class T>
+  inline int wordconcat(T word, int bstart, long int input, int bitsize) {
```
I suppose `T word` was supposed to be `T & word` , otherwise I can't see how it would work.

gpetruc commented on L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTSAMuonProducer.cc:
```diff
@@ -0,0 +1,196 @@
+  // We might not need to send BX info
+  //ap_int<BITSBX> bx(bx_);
+  int charge = mu.charge() > 0 ? 0 : 1;
+  ap_uint<BITSPT> pt = round(mu.pt() / 0.025);
```
and this should be `round(mu.pt() / 0.03125)` to match what's now in the interface document
